### PR TITLE
Re-enable collections

### LINKED ISSUE #785: NoMethodError: undefined method `remove_all' for #<Collection:0x000055bc78e208a8>
Sentry Issue: [CODEHARBOR-R](https://sentry.io/organizations/openhpi/issues/3531980598/?referrer=github_integration)

```
NoMethodError: undefined method `remove_all' for #<Collection:0x000055bc78e208a8>
  app/controllers/collections_controller.rb:54:in `remove_all'
    if @collection.remove_all
...
(83 additional frame(s) were not displayed)
```

### LINKED ISSUE #786: ActionView::Template::Error: undefined method `exercise_path' for #<ActionView::Base:0x00000000269f08>
Sentry Issue: [CODEHARBOR-P](https://sentry.io/organizations/openhpi/issues/3531936959/?referrer=github_integration)

```
NoMethodError: undefined method `exercise_path' for #<ActionView::Base:0x00000000269f08>
  app/views/collections/show.html.slim:34:
    = link_to exercise.title, exercise_path(exercise), class: 'btn btn-default disabled'
  app/views/collections/show.html.slim:28:
    - @collection.tasks.each do |exercise|
  app/controllers/collections_controller.rb:92:in `view_shared'
    render :show
...
(123 additional frame(s) were not displayed)

ActionView::Template::Error: undefined method `exercise_path' for #<ActionView::Base:0x00000000269f08>
```

### LINKED ISSUE #788: ActionView::Template::Error: undefined method `exercise_path' for #<ActionView::Base:0x00000000db8d28>
Sentry Issue: [CODEHARBOR-S](https://sentry.io/organizations/openhpi/issues/3559454302/?referrer=github_integration)

```
NoMethodError: undefined method `exercise_path' for #<ActionView::Base:0x00000000db8d28>
  app/views/collections/show.html.slim:32:
    = link_to exercise.title, exercise_path(exercise), class: 'btn btn-default'
  app/views/collections/show.html.slim:28:
    - @collection.tasks.each do |exercise|
...
(125 additional frame(s) were not displayed)

ActionView::Template::Error: undefined method `exercise_path' for #<ActionView::Base:0x00000000db8d28>
```

### LINKED ISSUE #809: ActiveRecord::InvalidForeignKey: PG::ForeignKeyViolation: ERROR:  update or delete on table "tasks" violates foreign key constrain...
Sentry Issue: [CODEHARBOR-W](https://sentry.io/organizations/openhpi/issues/3795029027/?referrer=github_integration)

```
PG::ForeignKeyViolation: ERROR:  update or delete on table "tasks" violates foreign key constraint "fk_rails_20b81c859c" on table "collection_tasks"
DETAIL:  Key (id)=(1) is still referenced from table "collection_tasks".

  app/controllers/tasks_controller.rb:58:in `destroy'
    @task.destroy!
...
(125 additional frame(s) were not displayed)

ActiveRecord::InvalidForeignKey: PG::ForeignKeyViolation: ERROR:  update or delete on table "tasks" violates foreign key constraint "fk_rails_20b81c859c" on table "collection_tasks"
DETAIL:  Key (id)=(1) is still referenced from table "collection_tasks".
```

### PR DESCRIPTION
**Pull request by Melhaya**
Fix removing one or all tasks if they are not in a collection already

Fixes #809, fixes #788, fixes #785, fixes #786

### CODE REVIEW

File: app/views/collections/show.html.slim [hunk 1]
```diff
@@ -25,19 +25,19 @@
...
-              - if can? :read, exercise
-                = link_to exercise.title, exercise_path(exercise), class: 'btn btn-default'
+              - if can? :read, task
+                = link_to task.title, task_path(task), class: 'btn btn-default'
               - else
-                = link_to exercise.title, exercise_path(exercise), class: 'btn btn-default disabled'
+                = link_to task.title, task_path(task), class: 'btn btn-default disabled'
             td
-              - if cannot? :read, exercise
+              - if cannot? :read, task
```
Comment (MrSerth) [inline app/views/collections/show.html.slim:36]: We should start commenting out the "contribute” action here.
```suggestion
              / Contributing to a task is not yet enabled again.
              /- if cannot? :read, task
```

File: app/views/collections/show.html.slim [hunk 2]
```diff
@@ -25,19 +25,19 @@
...
                 span.fa.fa-lock.lock
                 = ' '
-                = link_to t('collections.show.contribute'), contribute_exercise_path(exercise), method: :post, class: 'btn btn-default'
-              = link_to remove_exercise_collection_path(@collection, exercise: exercise), method: :patch, class: 'pull-right' do
+                /= link_to t('collections.show.contribute'), contribute_task_path(task), method: :post, class: 'btn btn-default'
```
Comment (MrSerth) [inline app/views/collections/show.html.slim:39]: As the previous commenting will also affect this line, we can revert the `/` here.

```suggestion
                = link_to t('collections.show.contribute'), contribute_task_path(task), method: :post, class: 'btn btn-default'
```

File: spec/controllers/collections_controller_spec.rb [hunk 3]
```diff
@@ -25,12 +25,20 @@
   end
 
   describe 'GET #show' do
-    let(:collection) { create(:collection, valid_attributes) }
+    render_views
```
Comment (MrSerth) [inline spec/controllers/collections_controller_spec.rb:28]: I would suggest adding a new line after `render_views`
```suggestion
    render_views
    
```

File: spec/controllers/collections_controller_spec.rb [hunk 4]
```diff
@@ -25,12 +25,20 @@
...
+    let(:task) { create(:task) }
+    let(:collection) { create(:collection, valid_attributes.merge(users: [user], tasks: [task])) }
 
     it 'assigns the requested collection as @collection' do
       get :show, params: {id: collection.to_param}
       expect(assigns(:collection)).to eq(collection)
     end
+
+    it 'includes a link to the respective tasks' do
+      # collection.tasks<<task
```
Comment (MrSerth) [inline spec/controllers/collections_controller_spec.rb:38]: This comment should be removed.

File: app/models/collection.rb [hunk 5]
```diff
@@ -14,13 +14,13 @@ class Collection < ApplicationRecord
   #   exercises << exercise unless exercises.find_by(id: exercise.id)
```
Comment (MrSerth) [inline app/models/collection.rb:14]: Could you please also fix the `add_exercise` method (and change its name to `add_task`)? Please, also remember to re-enable the tests.

It's currently not used, but we should change that as the last step. I'll add some references to the old code (which you might copy and adjust) or the references still included) below.

**Routes:**
https://github.com/openHPI/codeharbor/blob/2e08353bbbcf3f33369d602c0f1e6eae8c849d4a/config/routes.rb#L108

**Controller:**
https://github.com/openHPI/codeharbor/blob/2e08353bbbcf3f33369d602c0f1e6eae8c849d4a/app/controllers/exercises_controller.rb#L8
https://github.com/openHPI/codeharbor/blob/2e08353bbbcf3f33369d602c0f1e6eae8c849d4a/app/controllers/exercises_controller.rb#L124-L131

**View (still included, just need to uncomment PARTS):** 
`app/views/tasks/show.html.slim`, lines 513-523

**Localizations (just need an update):**
`config/locales/en.yml`, lines 93-94 + line 675

**Tests (Model):**
`spec/models/collection_spec.rb`, line 74ff.

**Tests (Controller):**
https://github.com/openHPI/codeharbor/blob/2e08353bbbcf3f33369d602c0f1e6eae8c849d4a/spec/controllers/exercises_controller_spec.rb#L268-L276

**Review by MrSerth — COMMENTED** [review]
Thanks! It's awesome to see the progress here and have collections almost restored completely. I just checked the code and think that adding a task to a collection is missing. So, I would suggest adding this as a last puzzle piece and then bring it live!

File: app/views/tasks/show.html.slim [hunk 6]
```diff
@@ -504,23 +504,23 @@ legend.toggle-next
...
     .button-description
-      = "Save Exercises for later by adding them to your cart or a collection."
+      = "Save Tasks for later by adding them to your cart or a collection."
     .wrapper
-      = link_to t('exercises.show.add_to_cart'), add_to_cart_exercise_path(@task), method: 'post', class: 'btn btn-default'
+      /  = link_to t('exercises.show.add_to_cart'), add_to_cart_exercise_path(@task), method: 'post', class: 'btn btn-default'
       .dropdown.btn-group
         = button_tag class: 'btn btn-default dropdown-toggle', data: {toggle: 'dropdown'} do
-          = t('exercises.show.add_to_collection') + ' '
+          = t('tasks.show.add_to_collection') + ' '
```
Comment (MrSerth) [inline app/views/tasks/show.html.slim:514]: This translation is missing.

File: app/views/tasks/show.html.slim [hunk 7]
```diff
@@ -504,23 +504,23 @@ legend.toggle-next
...
         ul.scrollable.dropdown-menu[role="menu"]
           - if !current_user.collections.empty?
             - current_user.collections.each do |collection|
               li
-                = link_to collection.title, add_to_collection_exercise_path(collection: collection), method: 'post', class: 'dropdown-item'
+                = link_to collection.title, add_to_collection_task_path(collection: collection), method: 'post', class: 'dropdown-item'
           - else
             li
-              = link_to t('exercises.show.add_collection'), new_collection_path(current_user)
+              = link_to t('tasks.show.add_collection'), new_collection_path(current_user)
```
Comment (MrSerth) [inline app/views/tasks/show.html.slim:523]: This translation is missing.

File: app/views/tasks/show.html.slim [hunk 8]
```diff
@@ -504,23 +504,23 @@ legend.toggle-next
                   .row-value
                     = file.usage_by_lms
 .button-panel
-  /.button-box
+  .button-box
     .button-description
-      = "Save Exercises for later by adding them to your cart or a collection."
+      = "Save Tasks for later by adding them to your cart or a collection."
```
Comment (MrSerth) [inline app/views/tasks/show.html.slim:509]: Would you mind moving this part to the en.yml, please? It should be part of the localizable strings, not just included in the HTML template.

File: app/views/tasks/show.html.slim (hunk 6 above)
Comment (MrSerth) [inline app/views/tasks/show.html.slim:514]: This is equal to (and for a better readability, we should probably switch):

```suggestion
          => t('tasks.show.add_to_collection')
```

[Read more here](https://github.com/slim-template/slim#output-).

**Review by MrSerth — COMMENTED** [review]
I just took the time to review your latest changes, please find my comments attached.

File: app/views/tasks/show.html.slim (hunk 8 above)
Comment (Melhaya) [reply app/views/tasks/show.html.slim:509]: Sure. I have added the corresponding translation in the en.yml file in the tasks.show section. Should I remove the corresponding ones from exercises.show?

Comment (MrSerth) [reply app/views/tasks/show.html.slim:509]: If it is not used somewhere else, yes please. If it is used somewhere else, please update the occurrences first and then proceed with removing it from the old path.

Comment (Melhaya) [reply app/views/tasks/show.html.slim:509]: Changes done and pushed :) Thanks for the feedback

Comment (MrSerth) [reply app/views/tasks/show.html.slim:509]: You're welcome -- thanks for incorporating it. Just two more things to go, I think, and then we should be quite ready 💪

**Comment by Melhaya** [comment]
The first 'add controller test' commit was a mistake. I think I didn't select any of the changes to be committed.

The second one has the right changes

**Comment by MrSerth** [comment]
> The first 'add controller test' commit was a mistake. I think I didn't select any of the changes to be committed.
> 
> The second one has the right changes

Thanks for making the latest changes! In case you made a mistake while committing, feel free to change your commit history afterward (i.e., by squashing or otherwise modify the history). This will make it much easier to review the changes and identify relevant changes later.

**Review by MrSerth — APPROVED** [review]
Thanks for working on re-enabling collections! Your work eliminates some existing bugs in the code base and restores more of the functionality of CodeHarbor. That's great news! 💪 

I just adjusted the commits and modified them, [please have a look at the commits](https://github.com/openHPI/codeharbor/pull/844/commits) (and my corresponding commit messages). I tried to regroup the changes, so that each controller action we touched is reflected in just one commit. That commit always contains everything to return to a green pipeline, so that it includes changes across the controller, model, views, and specs.

Additionally, while working on that, I made further changes. In order for view to view them, [here is a comparison](https://github.com/openHPI/codeharbor/compare/Melhaya-patch-2-original..Melhaya-patch-2). Please have a look there as well and try to understand a) what these changes do and b) why I did them. Feel free to post questions.

I will now proceed merging this PR, so that everyone can continue working with these latest changes 👍

**Pull request merged**